tucan/parser/parser.py

```diff
@@ -0,0 +1,164 @@
+
+    def _parse_sum_formula(self, formula_ctx):
+        if formula_ctx.getChildCount() == 0:
+            return
+
+        for symbol_count_tuple in formula_ctx.children:
+            symbol = symbol_count_tuple.getChild(0).getText()
+            count = 1
+            if symbol_count_tuple.getChildCount() > 1:
+                count = int(symbol_count_tuple.getChild(1).getText())
+            self._add_atoms(symbol, count)
+
+    def _add_atoms(self, element, count):
+        atom_props = {
+            "element_symbol": element,
+            "atomic_number": ELEMENT_PROPS[element]["atomic_number"],
+            "partition": 0,
+        }
+
+        for _ in range(count):
+            self._atoms.append(atom_props.copy())
+
+    def _add_bond(self, index1, index2):
+        self._bonds.append((index1 - 1, index2 - 1))
+
+    def _add_node_property(self, node_index, key, value):
+        props_for_node = self._node_attributes.setdefault(node_index - 1, {})
+
+        if key in props_for_node:
+            raise TucanParserException(
+                f'Atom {node_index}: Property "{key}" was already defined.'
+            )
+        props_for_node[key] = value
+
+    def to_graph(self) -> nx.Graph:
+        # node index validation
+        for i1, i2 in self._bonds:
+            self._validate_atom_index(i1)
+            self._validate_atom_index(i2)
+
+        sorted_atoms = sorted(self._atoms, key=lambda a: a["atomic_number"])
+
+        # dict of dict (atom_index -> dict of atom properties)
+        atoms_dict = {i: sorted_atoms[i] for i in range(len(sorted_atoms))}
+
+        # join in additional atom properties
+        for index, props in self._node_attributes.items():
+            self._validate_atom_index(index)
+
+            atom_props = atoms_dict[index]
+            atom_props.update(props)
+
+        # construct graph
+        graph = nx.Graph()
+        graph.add_nodes_from(list(atoms_dict.keys()))
+        nx.set_node_attributes(graph, atoms_dict)
+        graph.add_edges_from(self._bonds)
+
+        return graph
+
+    def _validate_atom_index(self, index):
+        if index >= len(self._atoms):
+            raise TucanParserException(f"Atom with index {index + 1} does not exist.")
+
+
+class RaisingErrorListener(ErrorListener):
+    def syntaxError(self, recognizer, offending_symbol, line, column, msg, e):
+        marked_error_location = self._underline_error(
+            recognizer, offending_symbol, line, column
+        )
+        error_str = f"line {line}:{column} {msg}\n{marked_error_location}"
+        raise TucanParserException(error_str)
+
+    def _underline_error(self, recognizer, offending_symbol, line, column):
+        pass
+
+
+# The algorithm of _underline_error was adopted from Terence Parr's book "The Definitive ANTLR 4 Reference", page 156.
+class LexerErrorListener(RaisingErrorListener):
+    def _underline_error(self, recognizer, offending_symbol, line, column):
```

Comment:
That's not so easy. :wink:
First of all, ANTRL forces me to use inheritance by extension of the `ErrorListener` class (part of antlr-runtime). Then, I want to have different behaviours in `_underline_error` for lexer and parser errors - another level of inheritance - and one of them doesn't use `offending_symbol`. The call of `_underline_error` in `RaisingErrorListener` enforces the method contract furthermore.
The overhead of not having the intermediate inheritance level (`RaisingErrorListener`) would be bigger, because there would be essentially the same `syntaxError` implementation for `LexerErrorListener` and `ParserErrorListener`

N.b.: Similar situation with the unused parameter `e` in `syntaxError`.

Reply:
Alright, looks like keeping the unused args seems to be the path of lesser resistance :wink: